src/execution/execute.js

```diff
@@ -977,6 +1022,47 @@ function completeAbstractValue(
 }
 
 /**
+ * If a resolveType function is not given, then a default resolve behavior is
+ * used which tests each possible type for the abstract type by calling
+ * isTypeOf for the object being coerced, returning the first type that matches.
+ */
+function resolveFirstValidType(
+  value: mixed,
+  context: mixed,
+  info: GraphQLResolveInfo,
+  possibleTypes: Array<GraphQLObjectType>,
+  i: number = 0,
```

Comment:
I'm a bit concerned about this function calling itself recursively. There are cases where a great many types can be iterated through here and the recursion could cause significant overhead. Also, if the matching type happens to be last in a long list of possible types, then it would have to wait for each promise to resolve sequentially before a match could be found.

I suggest calling the `isTypeOf` functions synchronously such that any promises generated run in parallel rather than in series. That way, if any synchronously return `true`, then the function may return synchronously, even if some of the `isTypeOf` functions return Promises. That also preserves the existing behavior. However if some responses were `thenable` then you could use `Promise.all` to resolve them in parallel and then iterate through the results to determine a match.